docs/design/2021-02-01-foreign-data-wrapper.md

```diff
@@ -0,0 +1,164 @@
+EXPLAIN SELECT city.cityName, people.name FROM city INNER JOIN people ON city.id=people.city
+> ID                 |    task  |
+  Project_0             root
+  |_ HashRightJoin      root
+    |- TableReader        root
+      \- TableScan        cop[tikv]
+  \_P 12                  root
+     \_TableReader        root
+        \_Selection       cop[csv]
+          \_TableScan     cop[csv]
+
+# 支持 custom string query，这里例如 TSDB
+SELECT * FROM TiDB_Metric where q=`avg(rate(TiDB_Index_Size{table=*,app='*'}[15m]) `;
+
+
+```
+
+在第一阶段 ( Version 1 ) ，着重实现 FDW 的核心功能。包括:
+
+- Create SERVER/ FOREIGN_TABLE
+
+- Scan Foreign Tables
+
+- Planning Post-Scan
+
+  例如 `SELECT * FROM table WHERE city="shanghai"`， 将 `"city=shanghai"` 下推到远程数据端做过滤
+
+- Insert Data to Foreign Table
+
+- EXPLAIN 支持 foreign table 
+
+## Rationale
+
+Mysql 通过内置的方式实现了多个 Engine，尚未有很好的流程支持远程数据存储。MySQL 官方只支持使用 MySQL 协议的远程数据存储。具有很大的局限性。
+
+PostgreSQL 制定 SQL/MED，实现了 FDW，通过 extension 的方式可以动态的开启相关 FDW 功能。
+
+Prometheus 自身 store 支持 [remote read/write](https://prometheus.io/docs/prometheus/latest/storage/#remote-storage-integrations)，proms 生态存在大量的围绕该功能对 proms 进行拓展的第三方库。
+
+TiDB 通过 Mysql 协议的兼容，实现了部分 engine API，但自身 Engine API 和自身生态 (TiKV/TiTan) 强绑定。无良好的插拔性。
+
+## Implementation
+
+###engine plugin
+
+```
+type EngineManifest struct {
+ Manifest
+ 
+ OnInsertOpen  func(ctx context.Context, meta *ExecutorMeta) error
+ OnInsertNext  func(ctx context.Context, rows [][]expression.Expression, meta *ExecutorMeta) error
+ OnInsertClose func(meta *ExecutorMeta) error
+
+ OnReaderOpen  func(ctx context.Context, meta *ExecutorMeta) error
+ OnReaderNext  func(ctx context.Context, chk *chunk.Chunk, meta *ExecutorMeta) error
+ OnReaderClose func(meta ExecutorMeta)
+
+ OnSelectReaderOpen func(ctx context.Context, filter []expression.Expression, meta *ExecutorMeta) error
+ OnSelectReaderNext func(ctx context.Context, chk *chunk.Chunk, filter []expression.Expression, meta *ExecutorMeta) error
+ OnSelectReaderClose func(meta ExecutorMeta)
+
+ OnCreateServer     func(sv *model.ServerInfo) error
+  OnDropServer     func(sv *model.ServerInfo) error
+
+ OnCreateTable      func(sv *model.ServerInfo, tb *model.TableInfo) error
+ OnDropTable        func(tb *model.TableInfo) error
+}
+
+```
+
+### 算子
+
+- `OnInsert` 用于对 Foreign Table 写入数据
+
+- `OnReader` 用于 scan foreign table
+
+- `OnDelete` 用于删除  foreign table 的数据
+
+- `OnUpdate` 用于更新 foreign table 中的数据
+
+- `OnSelect` 用于下推 scan 算子 
+
+### DDL
+
+- `OnCreateServer` 用于创建 Server, `CREATE SERVER {server_name}` 时被调用
+- `OnDropServer` 用于删除 Server, `Drop SERVER {server_name}` 时被调用
+- `OnCreateTable` 用于创建 table，`CREATE TABLE {table_name}`时被调用
+- `OnDropTable` 用于删除 table，`DROP TABKE {table_name}`时被调用
+
+## Compatibility and Limitations
+
+虽然本提案的计划只对 TiDB 主分支代码进行一些 Hook 埋点。作为一种对 Mysql SQL 的扩展，对已有功能未有任何修改与破坏，不存在兼容性问题。
+
+但因为以下原因，需要额外工作量：
+
+- 部分库和 Plugin 存在循环引用，需要针对性调整。
+- Mysql 的语法中并没有针对 `SQL/MED` 的设计，本提案遵循了 PostgreSQL 的语法与实现细节。同时将参考 MYSQL engine 的相关语法，针对性的进行 FDW 兼容。
```

Comment:
If there is no SQL/MED in MySQL, we might consider not using PG design but consider more of newer federated query engine like presto / spark or even dremio.